go/libraries/doltcore/remotestorage/chunk_fetcher.go

```diff
@@ -0,0 +1,586 @@
+	})
+	eg.Go(func() error {
+		return fetcherRPCDownloadLocsThread(ctx, locsReqCh, downloadLocCh, dcs.csClient, func(s string) { dcs.repoToken.Store(s) }, ret.resCh, dcs.host)
+	})
+	eg.Go(func() error {
+		return fetcherDownloadRangesThread(ctx, downloadLocCh, fetchReqCh, locDoneCh)
+	})
+	eg.Go(func() error {
+		return fetcherDownloadURLThreads(ctx, fetchReqCh, locDoneCh, ret.resCh, dcs.csClient, ret.stats, dcs.httpFetcher, dcs.params)
+	})
+
+	return ret
+}
+
+// Implements nbs.ChunkFetcher. Request the contents the chunks the given
+// |hashes|. They will be delivered through |Recv|. Returns an error if this
+// ChunkFetcher is terminally failed or if the supplied |ctx| is |Done|.
+func (f *ChunkFetcher) Get(ctx context.Context, hashes hash.HashSet) error {
+	select {
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	case <-f.egCtx.Done():
+		return context.Cause(f.egCtx)
+	case f.toGetCh <- hashes:
+		return nil
+	}
+}
+
+// Imeplements nbs.ChunkFetcher. Indicates that no further hashes will be
+// requested through |Get|. |Recv| will only return |io.EOF| after this is
+// called.
+func (f *ChunkFetcher) CloseSend() error {
+	close(f.toGetCh)
+	return nil
+}
+
+// Implements nbs.ChunkFetcher. Returns the next available
+// |nbs.CompressedChunk| whose contents have been fetched after being requested
+// by |Get|. Returns |io.EOF| after |CloseSend| is called and all requested
+// chunks have been successfully received. Returns an error if this
+// |ChunkFetcher| is terminally failed or if the supplied |ctx| is |Done|.
+func (f *ChunkFetcher) Recv(ctx context.Context) (nbs.CompressedChunk, error) {
+	select {
+	case <-ctx.Done():
+		return nbs.CompressedChunk{}, context.Cause(ctx)
+	case <-f.egCtx.Done():
+		return nbs.CompressedChunk{}, context.Cause(f.egCtx)
+	case cc, ok := <-f.resCh:
+		if !ok {
+			return nbs.CompressedChunk{}, io.EOF
+		}
+		return cc, nil
+	}
+}
+
+// Implements nbs.ChunkFetcher. Makes sure all resources associated with this
+// ChunkFetcher are released, and returns any errors encountered while fetching
+// requested chunks. This may return a non-|nil| error if the |ChunkFetcher| is
+// |Close|d before it has delivered |io.EOF| on a |Recv| call, but that is not
+// guaranteed. The correct way to guarantee everything has been received
+// without error is to read |Recv| until it returns |io.EOF|, and then to
+// |Close| the |ChunkFetcher|.
+func (f *ChunkFetcher) Close() error {
+	defer StatsFlusher(f.stats)
+	close(f.abortCh)
+	return f.eg.Wait()
+}
+
+// Reads HashSets from reqCh and batches all the received addresses
+// into |GetDownloadLocsRequest| messages with up to |batchSize| chunk hashes
+// in them. It delivers the batched messages to |resCh|.
+func fetcherHashSetToGetDlLocsReqsThread(ctx context.Context, reqCh chan hash.HashSet, abortCh chan struct{}, resCh chan *remotesapi.GetDownloadLocsRequest, batchSize int, repoPath string, idFunc func() (*remotesapi.RepoId, string)) error {
```

Comment:
The state machine in this function is pretty hard to follow. I'm not sure I understand it.

So `thisResCh` is assigned to `resCh` the first time there are any elements in `addr`. Prior to that, it's `nil.` Writing to a `nil` channel blocks forever. Is the idea that the select statement will never choose the `thisResCh <- thisRes` case until we assign `thisResCh` a non-nil value? Are we depending on the properties of writing to a `nil` channel for correctness here? Is this a good practice?

Maybe this is the idiomatic way to do this. But as someone less versed in go channels, the intended execution flow is very unclear.